File: JAVA/100 - Same Tree.java
/*
 100. Same Tree (E)

 Approach 1 : Preorder BFS
 Time: O(n)
 Space: O(p + q)

 Approach 2 : Recursion DFS
 Time: O(n)
 Space:O(n), n is the depth of the recursion

*/

/* Approach 1 */
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        Queue<TreeNode> todo = new LinkedList<TreeNode>();
        todo.add(p);
        todo.add(q);
        while (!todo.isEmpty()) {
            p = todo.poll();
            q = todo.poll();
            
            if (p == null && q == null) continue;
            if (p == null || q == null) return false;
            if (p.val != q.val) return false;
            
            todo.add(p.left);
            todo.add(q.left);
            todo.add(p.right);
            todo.add(q.right);
        }
        return true;
    }
}

/* Approach 2 */
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (p == null || q == null) return false;
        if (p.val != q.val) return false;
        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }
}

File: C++/100-Same_Tree.cpp
/*
 100. Same Tree (E)

 BFS
 Compare two nodes and push their children into queue, then pop out and 
 do the compare again till the end
 O(n),O(n)

 DFS
 O(n),O(n)

*/

// BFS
class Solution
{
public:
    bool isSameTree(TreeNode *p, TreeNode *q)
    {
        queue<TreeNode *> todo;
        todo.push(p);
        todo.push(q);
        while (!todo.empty())
        {
            p = todo.front();
            todo.pop();
            q = todo.front();
            todo.pop();

            if (!p && !q) continue;
            if (!p || !q) return false;
            if (p->val != q->val) return false;

            todo.push(p->left);
            todo.push(q->left);
            todo.push(p->right);
            todo.push(q->right);
        }
        return true;
    }
};

// DFS
class Solution
{
public:
    bool isSameTree(TreeNode *p, TreeNode *q)
    {
        if (!p && !q) return true;
        if (!p || !q) return false;
        if (p->val != q->val) return false;
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};